extensions/smallrye-health/deployment/src/test/java/io/quarkus/smallrye/health/test/HealthCheckContextPropagationTest.java

```diff
@@ -0,0 +1,78 @@
+package io.quarkus.smallrye.health.test;
+
+import static io.restassured.RestAssured.when;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.health.HealthCheck;
+import org.eclipse.microprofile.health.HealthCheckResponse;
+import org.eclipse.microprofile.health.Liveness;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.restassured.parsing.Parser;
+
+public class HealthCheckContextPropagationTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .overrideConfigKey("io.smallrye.health.context.propagation", "true")
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(RequestScopedBean.class, ContextualHC.class));
+
+    @Test
+    public void testContextPropagatedToHealthChecks() {
+        try {
+            RestAssured.defaultParser = Parser.JSON;
+
+            String firstResponse = when().get("/q/health").then()
+                    .body("status", is("UP"),
+                            "checks.status", contains("UP"),
+                            "checks.name", contains(ContextualHC.class.getName()),
+                            "checks.data", is(notNullValue()))
+                    .extract().response().jsonPath().getString("checks.data.request-scoped-instance");
+
+            String secondResponse = when().get("/q/health").then()
+                    .extract().response().jsonPath().getString("checks.data.request-scoped-instance");
+
+            String thirdResponse = when().get("/q/health").then()
+                    .extract().response().jsonPath().getString("checks.data.request-scoped-instance");
+
+            Assertions.assertFalse(firstResponse.equals(secondResponse) ||
+                    firstResponse.equals(thirdResponse) || secondResponse.equals(thirdResponse),
```

Comment:
This assertion looks unreliable, can we make it more objective?

Reply:
Not sure what you mean. Rewrite into separate assertions or is there a better way to check if 3 instances differ?

Reply:
Yeah, separate assertions would be better. My point is that it's not possible to know what condition triggered the assertion failure when you short-circuit conditions